pageserver/src/tenant.rs

```diff
@@ -112,34 +182,511 @@ pub struct Tenant {
+        }
+
+        info!("tenant {} attach complete", self.tenant_id);
+
+        // Spawn gc and compaction loops. The loops will shut themselves
+        // down when they notice that the tenant is inactive.
+        // TODO maybe use tokio::sync::watch instead?
+        crate::tenant_tasks::start_background_loops(self);
+
+        // We're ready for business.
+        // FIXME: Check if the state has changed to Stopping while we were downloading stuff
+        self.update_state(TenantState::Active);
+
+        for timeline in attached_timelines.iter() {
+            timeline.launch_wal_receiver()?;
+        }
+
+        Ok(())
+    }
+
+    ///
+    /// Load a tenant that's available on local disk
+    ///
+    /// This is used at pageserver startup, to rebuild the in-memory
+    /// structures from on-disk state. This is similar to attaching a tenant,
+    /// but the index files already exist on local disk, as well as some layer
+    /// files.
+    ///
+    /// If the loading fails for some reason, the Tenant will go into Broken
+    /// state.
+    ///
+    #[instrument(skip(conf, remote_storage), fields(tenant_id=%tenant_id))]
+    pub fn spawn_load(
+        conf: &'static PageServerConf,
+        tenant_id: TenantId,
+        remote_storage: Option<&GenericRemoteStorage>,
+    ) -> Result<Arc<Tenant>> {
+        // FIXME: also go into Broken state if this fails
+        let tenant_conf = Self::load_tenant_config(conf, tenant_id)?;
+
+        let wal_redo_manager = Arc::new(PostgresRedoManager::new(conf, tenant_id));
+        let tenant = Tenant::new(
+            TenantState::Loading,
+            conf,
+            tenant_conf,
+            wal_redo_manager,
+            tenant_id,
+            remote_storage.cloned(),
+        );
+        let tenant = Arc::new(tenant);
+
+        // Do all the hard work in a background task
+        let tenant_clone = Arc::clone(&tenant);
+
+        let _ = task_mgr::spawn(
+            &tokio::runtime::Handle::current(),
+            TaskKind::InitialLoad,
+            Some(tenant_id),
+            None,
+            "initial tenant load",
+            false,
+            async move {
+                match tenant_clone.load_tenant().await {
+                    Ok(()) => {}
+                    Err(err) => {
+                        tenant_clone.update_state(TenantState::Broken);
+                        error!("could not load tenant {tenant_id}: {err:?}");
+                    }
+                }
+                info!("initial load for tenant {tenant_id} finished!");
+                Ok(())
+            },
+        );
+
+        info!("spawned load of {} into background", tenant_id);
+
+        Ok(tenant)
+    }
+
+    ///
+    /// Background task to load in-memory data structures for this tenant, from
+    /// files on disk. Used at pageserver startup.
+    ///
+    #[instrument(skip(self), fields(tenant_id=%self.tenant_id))]
+    async fn load_tenant(self: &Arc<Tenant>) -> Result<()> {
```

Comment:
Asynchronously loading files into memory, and, generally, more async in tenant is definitely worth a standalone PR: as we've learned, async may have subtle bugs (cancellation, handler drops) that would need more attention that's hard to get in a massive PR.